### PR TITLE
docs: add a Distributed SQL section to the design doc

### PR DESCRIPTION
**Pull request by andreimatei**
cc @knz 

Radu, there was a bit of a push today to update the design doc, so I summarized the dist-sql rfc into a section.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/9657)

### CODE REVIEW

File: docs/design.md [hunk 1]
```diff
@@ -1310,6 +1310,265 @@ Finally, for SQL indexes, the KV key is formed using the SQL value of the
...
+language interpreter) to ease the exploitation of MapReduce. It provides a
+clear separation between "local" processes which process a limited amount of
+data and distributed computations, which are abstracted away behind a
+restricted set of conceptual constructs.
+
+To run SQL statements in a distributed fashion, we introduce a couple of concepts:
+- _logical plan_ - similar on the surface to the `planNode` tree described in
+  the [SQL](#sql) section, it represents the abstract data flow through
+  computation stages.
+- _physical plan_ - a physical plan is conceptuall a mapping of the _logical
```
Comment (knz) [inline docs/design.md]: typo

**Comment by RaduBerinde** [comment]
Thanks for doing this! I added a commit with some proposed fixes, nothing major, check it and maybe squash it if it's ok.

Also, the references don't seem to show up.

**Comment by tamird** [comment]
Seems a bit too detailed to me, so I stopped reading about halfway =/

---

Review status: 0 of 1 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[docs/design.md, line 1324 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657#-KSwgsu_JFZ8dC_jnKjV:-KSwgsu_JFZ8dC_jnKjW:b-iq121f) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2b51dcd4da54d878f9064c263680a829fbd3de2/docs/design.md#L1324)):*
> ```Markdown
> - Remote-side filtering: when querying for a set of rows that match a filtering
>   expression, instead of querying all the keys in certain ranges and processing
>   the filters after receiving the data on the gateway node over the network.
> ```

comma, not period?

---

*[docs/design.md, line 1330 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657#-KSwiJqs6vKyc4wtNt0N:-KSwiJqs6vKyc4wtNt0O:b432qjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2b51dcd4da54d878f9064c263680a829fbd3de2/docs/design.md#L1330)):*
> ```Markdown
>   perform the query and the updates on the node which has the data (as opposed
>   to receiving results at the gateway over the network, and then performing the
>   update or deletion there, which involves too many round-trips).
> ```

s/too many/additional/

---

*[docs/design.md, line 1332 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657#-KSwlUCdUJiprQA_3zH_:-KSwlUCdUJiprQA_3zHa:bjzhcdc) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cdbb70fd33a5e4f7453ef7e6559d3e2a582fed2/docs/design.md#L1332)):*
> ```Markdown
>   update or deletion there, which involves too many round-trips).
> - Parallelize SQL computation: when significant computation is required, we
>   want to distribute it too multiple node, so that it scales with the amount of
> ```

tooooooo

---

*[docs/design.md, line 1333 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657#-KSwlVtlwBmsWPzB8ADo:-KSwlVtlwBmsWPzB8ADp:b-artgug) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cdbb70fd33a5e4f7453ef7e6559d3e2a582fed2/docs/design.md#L1333)):*
> ```Markdown
> - Parallelize SQL computation: when significant computation is required, we
>   want to distribute it too multiple node, so that it scales with the amount of
>   data involved. This applies to `JOINs`, aggregation, sorting.
> ```

`JOIN`s

---

*[docs/design.md, line 1348 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657#-KSwp59Kc51Mwa9abSgN:-KSwp59Kc51Mwa9abSgO:b-tgfjpb) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cdbb70fd33a5e4f7453ef7e6559d3e2a582fed2/docs/design.md#L1348)):*
> ```Markdown
>   through computation stages.
> - _physical plan_ - a physical plan is conceptually a mapping of the _logical
>   plan_ nodes to CockroachDB nodes. Logical plan nodes are duplicated and
> ```

duplicated? what do you mean?

---

*[docs/design.md, line 1355 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657#-KSwpDvcGkgzRevdRL2V:-KSwpDvcGkgzRevdRL2W:b-nzw8hm) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cdbb70fd33a5e4f7453ef7e6559d3e2a582fed2/docs/design.md#L1355)):*
> ```Markdown
> 
>   The logical plan is made up of _aggregators_. Each _aggregator_ consumes an
>   _input stream_ of rows (or more streams for joins) and produces an _output
> ```

s/more/multiple/

---

*[docs/design.md, line 1356 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657#-KSwpTWXdshfPTw9-3al:-KSwpTWXdshfPTw9-3am:b-fl2n7w) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cdbb70fd33a5e4f7453ef7e6559d3e2a582fed2/docs/design.md#L1356)):*
> ```Markdown
>   The logical plan is made up of _aggregators_. Each _aggregator_ consumes an
>   _input stream_ of rows (or more streams for joins) and produces an _output
>   stream_ of rows. Both the input and the
> ```

weird wrapping

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657#-:-KSwrI5za7mMoLRQ6AGR:b-y9imtc)*

**Comment by a-robinson** [comment]
To second Tamir, this seems like a lot of detail for the top-level design doc, especially the explanations of all the aggregators and processors

---

Reviewed 1 of 1 files at r1.
Review status: 0 of 1 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657#-:-KT9qtYfLOrLlGHkRbTh:bs1sqly)*

**Comment by andreimatei** [comment]
I've cut down sentences here and there as I could. I don't think I can cut more... It's still long relative to other sections, but I'd say that its weight (and the weight of SQL in general relative to the whole doc) is too small, reflecting more the history of the document than what we should be communicating.
Anyway, stuff around "logical planning" is bound to change a lot as we start implementing it (currently we kind of worked around the whole thing), and that's the biggest section, so we're bound to rewrite most of this.

@raduberinde the references don't show up regardless of this change. I'm sending a separate PR to remove them altogether.

---

Review status: 0 of 1 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[docs/design.md, line 1324 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657#-KSwgsu_JFZ8dC_jnKjV:-K_5jYNo95zhPMrf6Yt_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2b51dcd4da54d878f9064c263680a829fbd3de2/docs/design.md#L1324)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

comma, not period?
</blockquote></details>

Done.

---

*[docs/design.md, line 1330 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657#-KSwiJqs6vKyc4wtNt0N:-K_5javD7CbnxcA8oPhi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2b51dcd4da54d878f9064c263680a829fbd3de2/docs/design.md#L1330)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

s/too many/additional/
</blockquote></details>

Done.

---

*[docs/design.md, line 1347 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657#-KSwfoC6RlMQjfgoqXLq-r1-1347:-K_5jiyGX-mNmEpFI0IQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2b51dcd4da54d878f9064c263680a829fbd3de2/docs/design.md#L1347)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

typo
</blockquote></details>

Done.

---

*[docs/design.md, line 1332 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657#-KSwlUCdUJiprQA_3zH_:-K_5jdJl-lQxcwZpmRqe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cdbb70fd33a5e4f7453ef7e6559d3e2a582fed2/docs/design.md#L1332)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

tooooooo
</blockquote></details>

Done.

---

*[docs/design.md, line 1333 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657#-KSwlVtlwBmsWPzB8ADo:-K_5rwtRtpgBc4T_tDvj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cdbb70fd33a5e4f7453ef7e6559d3e2a582fed2/docs/design.md#L1333)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

`JOIN`s
</blockquote></details>

Done.

---

*[docs/design.md, line 1348 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657#-KSwp59Kc51Mwa9abSgN:-K_5jjQrNuotU-TQ6f8q:b-ygf4gc) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cdbb70fd33a5e4f7453ef7e6559d3e2a582fed2/docs/design.md#L1348)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

duplicated? what do you mean?
</blockquote></details>

replaced with "replicated"

---

*[docs/design.md, line 1355 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657#-KSwpDvcGkgzRevdRL2V:-K_5jpaQOiZnX32WG0tE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cdbb70fd33a5e4f7453ef7e6559d3e2a582fed2/docs/design.md#L1355)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

s/more/multiple/
</blockquote></details>

Done.

---

*[docs/design.md, line 1356 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657#-KSwpTWXdshfPTw9-3al:-K_5jqYAVbNZpW2DZ_If:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cdbb70fd33a5e4f7453ef7e6559d3e2a582fed2/docs/design.md#L1356)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

weird wrapping
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657#-:-K_5rxzhoZ1pq7XPU6nX:b-6pmh7p)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[docs/design.md, line 1287 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657#-K_7F65KoXnyRQuT80wc:-K_7F65KoXnyRQuT80wd:bcknemy) ([raw file](https://github.com/cockroachdb/cockroach/blob/a48081d1a7b445a4b7b62de49b66dce7e28bf7f2/docs/design.md#L1287)):*
> ```Markdown
> goal of distributing the processing of SQL queries.
> See the [Distributed SQL
> RFC](https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/distributed_sql.md)
> ```

you should probably link to a stable SHA, or perhaps use a relative link (RFCS/distributed_sql.md).

---

*[docs/design.md, line 1328 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657#-K_7FIAELKt_P6-zDVou:-K_7FIAELKt_P6-zDVov:b-jxysq8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a48081d1a7b445a4b7b62de49b66dce7e28bf7f2/docs/design.md#L1328)):*
> ```Markdown
> streams are a logical concept and might not map to a single data stream in the
> actual computation. Aggregators will be potentially distributed when converting
> the *logical pla* to a *physical plan*; to express what distribution and
> ```

plan, not "pla"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657)*

**Comment by andreimatei** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[docs/design.md, line 1287 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657#-K_7F65KoXnyRQuT80wc:-K_a7LWAzv3UxLjsqsJp:b-2cfafm) ([raw file](https://github.com/cockroachdb/cockroach/blob/a48081d1a7b445a4b7b62de49b66dce7e28bf7f2/docs/design.md#L1287)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

you should probably link to a stable SHA, or perhaps use a relative link (RFCS/distributed_sql.md).
</blockquote></details>

I didn't use a stable SHA on purpose. The relative link makes sense, done.

---

*[docs/design.md, line 1328 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657#-K_7FIAELKt_P6-zDVou:-K_a7VaZwYZWueeLoS1U:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a48081d1a7b445a4b7b62de49b66dce7e28bf7f2/docs/design.md#L1328)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

plan, not "pla"
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657)*

**Comment by knz** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[docs/design.md, line 1398 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657#-K_aSTYX1YNcrs7eWxlZ:-K_aSTYYKDzKh8g2b7bR:b-uca8jb) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ee3d7d7c71e65b8641094fb663970bf66dec2c/docs/design.md#L1398)):*
> ```Markdown
>   connection to the client.
> 
> ## Physical planning
> ```

The transition between the two sections is now too difficult to read through: we're now missing the relationship between aggregators and processors. Could you add a few sentences to explain that aggregators are turned into processors during physical planning, and give an example mapping from aggregator to processor for a simple query.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657)*

**Comment by andreimatei** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[docs/design.md, line 1398 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657#-K_aSTYX1YNcrs7eWxlZ:-K_q0pv2WhIKHtpPTylP:b-6oenkn) ([raw file](https://github.com/cockroachdb/cockroach/blob/b03971f091e2d31f5e8a6f31cbd052e13bdadae3/docs/design.md#L1398)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

The transition between the two sections is now too difficult to read through: we're now missing the relationship between aggregators and processors. Could you add a few sentences to explain that aggregators are turned into processors during physical planning, and give an example mapping from aggregator to processor for a simple query.
</blockquote></details>

done. Added a few sentences and an example mapping in the Processors section below. The example is taken from the RFC and is kinda implanted here, but I think it's good enough to visualize things.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657)*

**Comment by knz** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[docs/design.md, line 1407 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657#-K_s4AoiCthzJrkGanfi:-K_s4AojyKNaKEG0pMuO:b-lf6whz) ([raw file](https://github.com/cockroachdb/cockroach/blob/f215e83661b84d4cc792f1fe5b7ecc40f724d2e1/docs/design.md#L1407)):*
> ```Markdown
> READER`s have multiple instances, split according to the ranges - each instance
> is planned on the lease holder of the relevant range. From that point on,
> subsequent processors are generally either collocated with their inputs, or
> ```

"colocated"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657#-:-K_s4GLT_YgYppSx-IDh:bnfp4nl)*

**Comment by andreimatei** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[docs/design.md, line 1407 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657#-K_s4AoiCthzJrkGanfi:-KaPP1mb2Mi0MWW6T8-N:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f215e83661b84d4cc792f1fe5b7ecc40f724d2e1/docs/design.md#L1407)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

"colocated"
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9657)*

**Pull request merged**